### PR TITLE
feat: add mechanical review-verdict computation gate to review.md

### PR DESCRIPTION
**Pull request by dodizzle**
## Summary
- Extracts the review.md Step 10 Case 1/Case 2 verdict truth table (documented as prose-only guidance) into a pure function `computeVerdict()` in a new `plugins/shipwright/scripts/compute-review-verdict.ts`, covering all 4 `selfReview` x `unaddressedFindings` combinations.
- Adds `validateReviewVerdict()`, a validation entrypoint that checks a constructed review body's `Verdict: ...` label against the computed expected label and returns a structured, actionable error on mismatch.
- Rewires review.md Step 10 to invoke the script for `event`/`verdictLabel` computation instead of freehand model judgment, and adds a new Step 10.5 that hard-gates on the validation result before the post-review step proceeds — a mismatch aborts and forces a fix.

## Acceptance Criteria
| # | Criterion | Status |
|---|-----------|--------|
| 1 | computeVerdict() covers all 4 truth-table combinations | MET |
| 2 | Validation entrypoint catches a mismatched label with a clear, actionable error | MET |
| 3 | review.md Step 10 rewritten to invoke the script and hard-gate on validation before posting | MET |
| 4 | New *.unit.test.ts covers all 4 combinations + mismatch-detection; no existing tests retired | MET |

## Test Plan
- `compute-review-verdict.unit.test.ts` — 11 tests covering all 4 truth-table rows, valid-match cases, the exact production mismatch (Case 1 mislabeled COMMENT), a reverse mismatch, missing-label detection, and casing/markdown edge cases.
- `review.content.test.ts` — 7 pre-existing tests re-pointed to the still-present `## Step 11: Post or Stage` boundary marker (only used as a search-window boundary, not an assertion target); all still pass.
- [x] `task lint`, `task typecheck`, `task ci` (coverage gate: 81.30% lines / 82.20% functions, threshold 80%) all pass. One pre-existing, unrelated failure in `agent/src/claude.unit.test.ts` confirmed present on `main` via `git stash` — not introduced by this change.

Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by dodizzle — COMMENTED** [review]
Verdict: COMMENT — computeVerdict() drops current-review-pass finding severity as an input, so an any-author PR with no prior unresolved GitHub review threads but a fresh critical/important finding in this pass gets silently computed as APPROVE. This contradicts unchanged text in the same file (Step 9.5's own fallback formula, the "APPROVE means clean" Review Quality Rule, and references/post-review-guide.md's APPROVE vs COMMENT Rule). See inline comment for details.

File: plugins/shipwright/scripts/compute-review-verdict.ts [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+// `event` is COMMENT whenever selfReview is true (GitHub rejects self-APPROVE
+// via the API) OR unaddressedFindings is true (Step 9.5's hard gate) — either
+// condition alone is sufficient, matching Step 9.5/Step 10's "not mutually
+// exclusive" note. `verdictLabel` reflects the *actual* quality verdict:
+// COMMENT only when there is a genuine unaddressed finding, regardless of why
+// `event` was forced to COMMENT.
+export function computeVerdict(
+  input: ComputeVerdictInput,
+): ComputeVerdictResult {
+  const event: Verdict =
```
Comment (dodizzle) [inline plugins/shipwright/scripts/compute-review-verdict.ts:76]: This function's only inputs are `selfReview` and `unaddressedFindings` (Step 9.5's definition — unresolved *prior* GitHub review threads/comments only). It has no path to reflect a critical/important finding the code-reviewer subagent finds fresh in *this* review pass (Steps 7/8) when there are zero prior unresolved threads.

The old "Event selection" prose this PR deletes from review.md Step 10 read: "If any finding at important/critical severity remains after threshold filtering: COMMENT — no exceptions." That logic has no replacement here.

This contradicts text left unchanged in the same file:
- Step 9.5 itself (review.md:547-549) already documents the correct fallback formula — `unaddressedFindings ? "COMMENT" : {code-reviewer subagent's recommended verdict}` — implying the non-unaddressed-findings branch should still consult the subagent's recommendation, not hardcode APPROVE.
- The "Review Quality Rules" section (review.md:1047-1048): "APPROVE means clean: any finding at important or critical severity means COMMENT."
- references/post-review-guide.md's "APPROVE vs COMMENT Rule", which review.md Step 10 still points readers to — now stale, describing behavior this script doesn't implement.

Net effect: an any-author PR with no prior unresolved GitHub review threads but a fresh critical finding in this pass gets `event:"APPROVE", verdictLabel:"APPROVE"`, and Step 10.5's hard-gate validation reports `valid:true` (the label matches the wrongly-computed expected value) — silently posting a clean APPROVE on a PR with a blocking bug. This is the same failure mode the PR's own "Worked example" says it's fixing, reintroduced by the fix itself.

Suggestion: add a third boolean input (e.g. `currentPassHasBlockingFindings`, derived from Step 8's threshold-filtered findings) and OR it into both the `event`-COMMENT and `verdictLabel`-COMMENT branches to restore the old behavior. Update the truth table to 8 rows and add a covering test. Also update references/post-review-guide.md's APPROVE vs COMMENT Rule to match.

File: plugins/shipwright/commands/review.md [hunk 2]
```diff
@@ -593,35 +619,65 @@ label, on both the initial-review and re-review paths (Steps 10/11 run identical
...
-authorship vs. a real quality gate on content) — the `Verdict: ...` body label is the *only*
-signal that distinguishes them for downstream automation, so it must always reflect the actual
-computed verdict, never be copied from the `event` value or from generic "this went through
-COMMENT" boilerplate.
+  gate computes real unaddressed findings (`unaddressedFindings = true`) — an unresolved inline
+  thread, or a qualifying `COMMENTED`/`CHANGES_REQUESTED` review body not excluded by the
+  clean-APPROVE or author-reply rules. This holds regardless of `selfReview` or
+  `currentPassHasBlockingFindings`. Result: `event: "COMMENT"`, body correctly reads
+  `"Verdict: COMMENT — ..."`.
+- **Case 3 — any-author PR, no prior unaddressed findings, but a fresh blocking finding in
```
Comment (dodizzle) [inline plugins/shipwright/commands/review.md:663]: This PR adds substantial new prose here (Case 3, the 8-row truth table, Step 8's `CURRENT_PASS_HAS_BLOCKING_FINDINGS` instruction), but `review.content.test.ts` — the designated content-assertion layer for exactly this kind of markdown-body verification — wasn't extended to cover it. The existing RVG-1.2 `describe` block already has dedicated assertions for Case 1 and Case 2 wording; this PR's changes to that file are limited to six mechanical search-boundary-string replacements (`**Event selection**` → `## Step 11: Post or Stage`), with no new `it()` blocks for Case 3. A future edit could silently corrupt or delete this new prose without any content test catching it. Suggest adding assertions mirroring the existing Case 1/Case 2 pattern for Case 3, `currentPassHasBlockingFindings`, and the 8-row table header.

**Review by dodizzle — COMMENTED** [review]
Verdict: COMMENT — The prior critical finding (computeVerdict() dropping current-pass finding severity) is correctly fixed: the new currentPassHasBlockingFindings input restores the old "any important/critical finding forces COMMENT" behavior across all 8 truth-table rows, and it's fully unit-tested. One new gap this pass: review.content.test.ts wasn't extended to cover the new Case 3 / 8-row-truth-table prose it introduces — see inline comment.

File: plugins/shipwright/scripts/compute-review-verdict.ts [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+//
+// Mirrors agent/src/check-helpers.ts's VERDICT_APPROVE_LABEL convention
+// (`/verdict\**\s*:\s*\**approve\b/i`) — case-insensitive, optional markdown
+// bold markers around "verdict", the colon, and/or the verdict word, not
+// anchored to end-of-line since review bodies trail reasoning after the
+// verdict on the same line. Duplicated here (rather than imported) because
+// plugins/shipwright is a separate, repo-agnostic package from agent/src —
+// see plugins/shipwright/CLAUDE.md.
+
+const VERDICT_APPROVE_LABEL = /verdict\**\s*:\**\s*\**approve\b/i;
```
Comment (dodizzle) [inline plugins/shipwright/scripts/compute-review-verdict.ts]: This regex diverges from the canonical `agent/src/check-helpers.ts:119` pattern it claims to mirror. Canonical: `/verdict\**\s*:\s*\**approve\b/i`. Here: `/verdict\**\s*:\**\s*\**approve\b/i` — an extra `\**` right after the colon.

Verified empirically: on `"**Verdict:** approve — all checks pass."` (the exact body used in this file's own unit test a few lines below, "tolerates markdown bold markers around the label"), the canonical `check-helpers.ts` regex returns `false` while this regex returns `true`. So `validateReviewVerdict()` accepts a body here that `agent/src/check-patch.ts`'s `isSelfCleanApprove` (which uses the canonical pattern) would NOT recognize as a clean self-approve downstream — reintroducing this PR's own motivating bug via a different code path.

Suggest changing both regexes here to match `check-helpers.ts` exactly (`/verdict\**\s*:\s*\**approve\b/i` / `.../comment\b/i`). Same issue applies to `VERDICT_COMMENT_LABEL` on the next line.

File: plugins/shipwright/commands/review.md [hunk 4]
```diff
@@ -571,20 +579,38 @@ store.
...
+  THIS review pass contains at least one `important` or `critical` severity finding, else
+  `false`. This is distinct from `unaddressedFindings`: it reflects what the code-reviewer
+  subagent found fresh in Steps 7/8, not prior GitHub-posted review state.
+
+Invoke `compute-review-verdict.ts` to turn those three booleans into the `event` and
+`verdictLabel` to use — do not decide the event or the body's `Verdict: ...` label by
+narrative judgment:
+
+```bash
+bun run "${CLAUDE_PLUGIN_ROOT}/scripts/compute-review-verdict.ts" \
```
Comment (dodizzle) [inline plugins/shipwright/commands/review.md:612]: This fence interpolates `${SELF_REVIEW}`, `${UNADDRESSED_FINDINGS}`, `${CURRENT_PASS_HAS_BLOCKING_FINDINGS}` — none of these UPPER_SNAKE_CASE shell variables is ever assigned anywhere in review.md. The narrative a few lines above only defines camelCase concepts (`selfReview`, `unaddressedFindings`, `currentPassHasBlockingFindings`), never `SELF_REVIEW=...` etc.

This doc is a literal runbook the review agent executes — a literal run of this fence would interpolate empty strings (or error under `set -u`), producing invalid JSON passed to `compute-review-verdict.ts`. The same three vars plus `$BODY` are also unassigned in Step 10.5's fence (line 717-718). Suggest adding explicit assignment steps before each fence, or rewriting the fences as illustrative pseudo-code rather than literal `${VAR}` interpolation.

**Review by dodizzle — COMMENTED** [review]
Verdict: COMMENT — the two prior findings (severity-gate regression, missing content-test coverage) are correctly fixed and verified. This pass found two new important-severity gaps: the duplicated VERDICT_APPROVE_LABEL/VERDICT_COMMENT_LABEL regex in compute-review-verdict.ts diverges from its canonical agent/src/check-helpers.ts source (verified empirically — they disagree on this PR's own test input), and Step 10/10.5's bash fences reference SELF_REVIEW/UNADDRESSED_FINDINGS/CURRENT_PASS_HAS_BLOCKING_FINDINGS/BODY shell variables that are never assigned anywhere in review.md. See inline comments.

**Review by dodizzle — COMMENTED** [review]
Verdict: COMMENT — Both prior findings (regex divergence from the canonical check-helpers.ts pattern, unassigned shell variables in the Step 10/10.5 bash fences) are correctly fixed and independently verified this round. This pass surfaces two new issues, both confined to Step 9.5 (not touched by this PR, but sitting immediately upstream of Step 10 in the same walked procedure):

1. **Step 9.5's verdict formula is stale and contradicts Step 10's new mechanical truth table** (`review.md:555-556`, confidence 90). Step 9.5 still reads `compute the verdict once (unaddressedFindings ? "COMMENT" : {code-reviewer subagent's recommended verdict})` — a two-input formula that falls back to the subagent's own top-line `recommendation` whenever `unaddressedFindings` is false. Step 10's new truth table says the verdict must also be COMMENT when `currentPassHasBlockingFindings` is true (Step 8's threshold-filtered findings, independent of the subagent's own `recommendation` field). This is exactly the freehand-fallback path Case 3 of the new worked example was written to eliminate — a direct self-contradiction undermining this PR's own "mechanical not freehand" goal. Suggest updating Step 9.5's parenthetical to reference the full three-input `compute-review-verdict.ts` call, or drop the inline formula and point to Step 10.

2. **Step 9.5's cross-reference to the "Step 10 self-review COMMENT-forcing override" now points to a deleted subsection** (`review.md:564`, confidence 80). This PR removes the named "Self-review event override" subsection and folds it into the new unnamed mechanical paragraph, so the cross-reference no longer resolves. Suggest pointing it at the new mechanical paragraph/truth table instead (e.g. "the `selfReview` input to Step 10's mechanical computation").

Both findings were independently confirmed by a dispatched code-reviewer subagent pass. Strong PR overall — the mechanical verdict computation, hard-gate validation, and new test coverage are real improvements that close a recurring production bug; these two remaining issues are localized documentation-consistency gaps in an adjacent, unmodified section.

**Review by dodizzle — COMMENTED** [review]
Verdict: COMMENT — Both prior findings (Step 9.5's stale verdict formula and dangling cross-reference to a deleted subsection) are correctly fixed and verified in review.md. This pass found one new issue: `plugins/shipwright/commands/patch.md:240` still cites the "Self-review event override" subsection name that this PR's Step 10 rewrite deletes from review.md — a dangling cross-reference in a sibling file the diff doesn't touch, so review.content.test.ts's review.md-scoped assertions can't catch it. Suggest updating that citation to reference the mechanical computation generically instead of the now-deleted subsection name. The core mechanical verdict-computation work (8-row truth table, compute-review-verdict.ts, Step 10.5 hard-gate validation, and test coverage across all 8 rows) is solid.

**Review by dodizzle — COMMENTED** [review]
Verdict: COMMENT — No new issues found this round. The round-5 fix (patch.md:240's dangling cross-reference to the deleted "Self-review event override" subsection) is verified correct, and an independent fresh review of the full diff found zero remaining or new issues; all lint/typecheck-relevant tests pass. Posted as COMMENT rather than APPROVE: this is a self-authored PR (GitHub blocks self-APPROVE via the API), and Step 9.5's unaddressed-findings hard gate still counts the prior five COMMENT review passes as unaddressed (none is a clean-APPROVE body, and no PR-level author-reply comment exists to trigger the reply exclusion, since this PR was reviewed exclusively via new review objects each round).

**Review by dmcaulay — COMMENTED** [review]
Verdict: COMMENT — No new issues found in round 7 or across the full diff (tests pass, lint clean, computeVerdict()'s 8-row truth table verified consistent with review.md's worked example and check-helpers.ts's canonical regex). Forced to COMMENT by the currently-deployed review.md's Step 9.5 gate, which counts this PR's five prior COMMENTED reviews (no clean-APPROVE body, no qualifying author-reply) as unaddressed findings regardless of this pass's own findings — the self-referential quirk this PR fixes, effective once merged.

**Comment by dodizzle** [comment]
Reviewed all 7 rounds of findings on this PR against the current code in the worktree (HEAD `e4cec2bb`, which matches this PR's current head) and confirmed none requires further code changes:

- Round 1 (`computeVerdict()` silently dropping current-pass finding severity): fixed — `currentPassHasBlockingFindings` is a documented input in `compute-review-verdict.ts`'s 8-row truth table and forces `COMMENT` whenever true, independent of `unaddressedFindings`.
- Round 2 (missing `review.content.test.ts` coverage for Case 3 / the 8-row truth table): fixed — the file now has dedicated tests for Case 3 (`"Case 3 describes any-author PR..."`, `"...same failure mode as Case 1"`) and asserts the full 8-row truth table's event/verdictLabel outputs.
- Round 3 (`VERDICT_APPROVE_LABEL`/`VERDICT_COMMENT_LABEL` regex divergence from `agent/src/check-helpers.ts`, plus unassigned `SELF_REVIEW`/`UNADDRESSED_FINDINGS`/`CURRENT_PASS_HAS_BLOCKING_FINDINGS`/`BODY` shell vars in review.md's Step 10/10.5 fences): fixed — `compute-review-verdict.ts`'s regexes are byte-identical to `check-helpers.ts`'s (`/verdict\**\s*:\s*\**approve\b/i` and the COMMENT equivalent), and review.md:606-608 and :733 now assign each shell var from the corresponding computed value before the `bun run compute-review-verdict.ts` invocation.
- Round 4 (Step 9.5's stale two-input verdict formula, plus a dangling cross-reference to a deleted "Step 10 self-review COMMENT-forcing override" subsection): fixed — Step 9.5 now explicitly defers to Step 10's full three-input `compute-review-verdict.ts` call rather than restating an inline formula, and its cross-reference points at the new mechanical truth-table paragraph instead of the removed subsection name.
- Round 5 (patch.md:240 citing the deleted "Self-review event override" subsection name): fixed — patch.md's Step 3a now describes the exclusions generically (clean self-APPROVE, author-reply, and the newer superseded-self-review case) without citing the removed subsection name.
- Round 6 & 7 (no new issues found): confirmed no outstanding code-level finding. Both rounds were mechanically forced to COMMENT by the currently-deployed (pre-merge) review.md's own Step 9.5 gate — the exact self-referential bug this PR fixes — not by any problem in this PR's code. Notably, a prior patch cycle already went further and added a DRO-1.2 "superseded-self-review" exclusion to patch.md/review.md specifically to resolve this self-referential quirk for future rounds once merged.

Validation: `task lint` is clean, and both PR-relevant test files (`review.content.test.ts`, `compute-review-verdict.unit.test.ts`) pass fully (102/102). The broader `task test` run has unrelated pre-existing failures (e.g. missing `@hono/zod-openapi`/`@shipwright/lib` module resolution in task-store/metrics packages, sandboxed network/auth errors in agent integration tests) that reproduce identically on `main` and are unrelated to this PR's diff.

Posting this comment so a subsequent review can reach a clean APPROVE per the reply-exclusion this PR itself introduces.

File: plugins/shipwright/commands/patch.md [hunk 5]
```diff
@@ -266,9 +269,32 @@ self-authored review. This exclusion still requires all inline threads to be res
 (`isResolved == true`) — an unresolved thread on the same review continues to count as a
 finding regardless of any reply.
 
+**Self-review superseded by a later clean self-review (DRO-1.2)**: review.md's own Step
```
Comment (dmcaulay) [inline plugins/shipwright/commands/patch.md:272]: This new "superseded self-review" exclusion (DRO-1.2) isn't mirrored in `agent/src/check-patch.ts`'s `hasUnaddressedFindings()` — that function only implements `isSelfCleanApprove` (CPF-2.1) and `isAddressedByAuthorReply` (CPF-2.3), no DRO-1.2 equivalent. Per the plugin's Design Constitution, the `agent/src` candidate providers are authoritative on what qualifies for autonomous dispatch — `getPatchCandidates()` calls `hasUnaddressedFindings()` directly, not this prose. So a self-authored PR stuck in the "N-1 stale COMMENT self-reviews" pattern (this PR's own motivating case) still won't be surfaced as a `patch` candidate by the `shipwright-loop` cron, even though a direct manual `/shipwright:patch org/repo#N` invocation would now correctly exclude it. Worth mirroring this exclusion in `check-patch.ts` (or filing a tracked follow-up) so the loop-driven path is actually fixed, not just the manual-invocation path.

**Review by dmcaulay — COMMENTED** [review]
Verdict: COMMENT — check-patch.ts's hasUnaddressedFindings() doesn't mirror the new DRO-1.2 "superseded self-review" exclusion this PR adds to patch.md's Step 3a, so the autonomous shipwright-loop patch-candidate path still won't recognize a self-authored PR's stale prior COMMENT self-reviews as resolved — only a direct manual /shipwright:patch invocation picks up the new prose. Everything else in this pass checks out: tests pass (102/102), lint clean, computeVerdict()'s 8-row truth table and regex byte-match verified independently, and the prior 7 rounds' findings are now addressed per the author's reply (unaddressedFindings=false), but this fresh finding forces COMMENT (currentPassHasBlockingFindings=true — this PR's own Case 3).

**Review by dmcaulay — COMMENTED** [review]
Verdict: COMMENT — Round 8's finding (check-patch.ts's hasUnaddressedFindings() missing the DRO-1.2 superseded-self-review mirror) is fixed and verified: isSupersededBySelfReview() is correctly scoped, with 4 new unit tests covering the superseded case, the non-clean-later-review case, order-matters, and the third-party-not-superseded case. No new issues found this pass — tests (166/166), lint, and CI are all clean, and a second independent review pass also found zero findings. Posting as COMMENT rather than APPROVE because the currently-deployed Step 9.5 gate still counts round 8's own review as an unaddressed finding (no author reply landed after it) — a self-referential quirk of re-reviewing this PR's own history that only resolves once merged, not a defect in the diff.

**Comment by dodizzle** [comment]
## Patch pass: both open review findings confirmed already resolved in code

I re-examined the current branch HEAD against rounds 8 and 9. Both findings are already addressed — no further code change is needed.

### Round 8 (dmcaulay, 2026-08-09T10:34:35Z) — RESOLVED IN CODE

The finding: `check-patch.ts`'s `hasUnaddressedFindings()` did not mirror the new DRO-1.2 "superseded self-review" exclusion that this PR adds to `patch.md`'s Step 3a.

This was fixed in commit `b5cbbb0` ("fix: address review findings on #2515 — mirror DRO-1.2 superseded-self-review exclusion in check-patch.ts", 2026-08-09T10:45:14Z), which postdates the round-8 review. Verified in the current worktree:

- **`isSupersededBySelfReview(review, allReviews, currentUser)`** — `agent/src/check-patch.ts` (lines 274-288). Correctly scoped: returns `false` for non-self reviews (`review.author.login !== currentUser`), and only supersedes a prior self-review when a **later** self-review (`submittedAt` strictly greater, same `author.login`) is itself a clean verdict per `isCleanApproveBody`.
- It is wired into the qualifying-reviews filter in `hasUnaddressedFindings()` (line 322): `!isSupersededBySelfReview(r, reviews.nodes, currentUser)`.
- The logic matches `patch.md` Step 3a's "Self-review superseded by a later clean self-review (DRO-1.2)" prose (lines 272-293) one-to-one: only a prior self-review is superseded, only when the later self-review is genuinely clean; a later non-clean self-review supersedes nothing; third-party findings are untouched; inline threads must still be resolved.

Test coverage (4 tests, `agent/src/check-patch.unit.test.ts`, lines 985-1140):
- "returns empty array when an earlier self-authored COMMENT review is superseded by a later clean self-review" (the superseded case)
- "returns a candidate when an earlier self-authored COMMENT review is followed by a LATER self-review that is itself non-clean" (non-clean-later-review case)
- "does NOT supersede an earlier self-review when the clean self-review comes BEFORE it (order matters)" (order-matters case)
- "does NOT supersede a THIRD-PARTY review's finding, even when a later self-review is clean" (third-party-not-superseded case)

### Round 9 (dmcaulay, 2026-08-09T10:57:47Z) — CONFIRMED, no code change required

Round 9 itself states round 8 is "fixed and verified" and that "No new issues found this pass" — it is not a defect in the diff. It explains that COMMENT (rather than APPROVE) is solely a self-referential quirk of the currently-deployed Step 9.5 gate counting round 8's own review as unaddressed until this PR merges. No action is possible or needed here.

### Validation

- `bun test agent/src/check-patch.unit.test.ts` — 64/64 pass
- `bun test plugins/shipwright/scripts/compute-review-verdict.unit.test.ts plugins/shipwright/commands/review.content.test.ts` — 102/102 pass (166 total across this PR's changed files, matching round 9's count)
- `task lint` — clean

No files changed this pass; both findings were already resolved in branch history.

**Review by dmcaulay — APPROVED** [review]
Verdict: APPROVE — Round 10: all three verdict inputs are now clean. Round 8's fresh finding (DRO-1.2 exclusion not mirrored in check-patch.ts) was fixed and verified in b5cbbb05 with 4 new unit tests. The PR author's follow-up comment now post-dates every prior review on this PR, so `unaddressedFindings` correctly resolves to false via the author-reply exclusion. An independent code-reviewer pass over the full diff (computeVerdict()'s 8-row truth table, isSupersededBySelfReview()'s edge cases, regex parity with check-helpers.ts, dangling cross-references, test coverage) found zero findings. Tests pass (166/166), CI is green at this commit.
